### ISSUE TITLE
ProhibitMagicNumbers suggested split into array offset and not

### ISSUE THREAD

**Issue by epa** [Perl-Critic/Perl-Critic#621]
ProhibitMagicNumbers is a useful policy.  It is also warns about array accesses with odd offsets, as

````
    my $value = $a[5];
    my $size = $a[6];
````

These hardcoded constants 5 and 6 can be considered a code smell and suggest rewriting the code to use a hash with named fields instead.

And yet... code that passes around tuples and unpicks them by accessing particular numbered columns is quite common.  Although the magic numbers 5 and 6 for array offsets are not great, they are not quite as bad as having them as magic numbers in other contexts.  It is, at least, fairly self-explanatory that 5 is the value column and 6 is the size column; the code would not be made any easier to understand by adding boilerplate constants VALUE_COLUMN_OFFSET and SIZE_COLUMN_OFFSET.  (Admittedly those will still help if the same structure is used in several places in the code and the programmer wants to change the order of elements.  But that applies too to offsets 0 and 1, which the policy won't warn about.)

Therefore I suggest the policy might be split into two: ProhibitMagicNumbersAsArrayOffset, and ProhibitMagicNumbersNotArrayOffset.  Sorry these suggested names are rather a mouthful.  The idea is that the policy against magic numbers outside array offsets could be enabled by default at a lower severity level, while the restriction on $a[5] would only be enforced at brutal level.  And of course users would be able to turn on and off the two policies separately.

**Comment by petdance** [comment]
Agreed on the premise, but it would be far simpler to have how about making it a config option rather two entirely different policies?

    [ValuesAndExpressions::ProhibitMagicNumbers]
    allow_indexes = yes

**Comment by epa** [comment]
A config option would also work.  I suggested the split because the idea of having different severity levels sounded nice.  Perhaps at some point configuration options could be tied to severity levels.

````
[severity 3, 4, 5] allow_indexes = yes
[severity 1, 2] allow_indexes = no
````

But yes, for the scope of this bug, a boolean config option would be fine.

**Comment by petdance** [comment]
index_severity = 3
    nonindex_severity = 5

**Comment by thaljef** [comment]
Giving each variant a different severity is doable. You just have to construct the `Violation` objects directly and pass in the severity, rather than using the convenience method that interrogates your policy for its severity.

I'm on board with adding configuration to make indexes exempt. But I'm not convinced that giving them separate severities is really warranted.

**Comment by epa** [comment]
Yes, the idea of different severities just appeals to my sense of tidyness.  It probably doesn't achieve much because I doubt that many people use the severities - except when first auditing a codebase.  I imagine most do what I do, run with max severity and then disable policies they don't like.

As long as the error message hints about what to do: 'since this is an array index, you can turn off the check with allow_indexes=yes...' that would work fine.  Two separate policies are clunky, but at least it is obvious how to turn one of them off.
